zio-kafka/src/main/scala/zio/kafka/consumer/internal/PartitionStreamControl.scala

```diff
@@ -108,17 +112,26 @@ final class PartitionStreamControl private (
   private[internal] def end: ZIO[Any, Nothing, Unit] =
     logAnnotate {
       ZIO.logDebug(s"Partition ${tp.toString} ending") *>
-        dataQueue.offer(Take.end).unit
+        dataQueue.offer(Take.end).unit *> hasEndedRef.set(true)
```

Comment:
Shouldn't this be the other way around? With the original order method `offerRecords` could insert new records after `offer(Take.end)` and before `hasEndedRef.set(true)`.

```suggestion
        hasEndedRef.set(true) *>
        dataQueue.offer(Take.end).unit
```